File: bin/repl.py
#!/usr/bin/env python
from src import parse
from src import evaluation
from src import main
from src import runtime


def _read():
    return raw_input(' $ ')


def _eval(string, env):
    return evaluation.evaluate(parse.Parser(string).parse_expression(), env)


def _print(string):
    print(str(string))


def _loop(env=main.DefaultEnvironment()):
    while True:
        evaluation = _eval(_read(), env)
        if isinstance(evaluation, runtime.BindingType):
            env = evaluation.env
        _print(evaluation)


if __name__ == '__main__':
    _loop()

File: haskell/Repl.hs
import qualified Parse
import qualified Evaluation
import qualified Runtime
import qualified Env


read_ :: IO String
read_ = getLine

loop env = do
    command <- getLine
    print_ $ eval command env
    loop env

eval :: String -> Runtime.Environment -> String
eval command env = Runtime.value $ Evaluation.evaluate (Parse.parse command) env

print_ :: String -> IO ()
print_ string = do putStrLn $ string


main = do
    putStrLn "Welcome to the Fresh Interpreter!"
    putStrLn "(+ 3 4)"
    putStrLn "7"
    loop Env.defaultEnvironment
